libafl/src/executors/forkserver.rs

```diff
@@ -0,0 +1,452 @@
+
+
+        //forkserver
+        let forkserver = Forkserver::new(target.clone(), args.clone(), out_file.as_raw_fd(), use_stdin, 0);
+        let (rlen, _) = forkserver.read_st();//initial handshake
+        if rlen == 4{
+            println!("Forkserver up!!");
+        }
+
+        Ok(Self {
+            target,
+            args,
+            use_stdin,
+            out_file,
+            forkserver,
+            observers,
+            phantom: PhantomData,
+        })
+    }
+
+    pub fn target(&self) -> &String {
+        &self.target
+    }
+
+    pub fn args(&self) -> &Vec<String>{
+        &self.args
+    }
+
+    pub fn forkserver(&self) -> &Forkserver{
+        &self.forkserver
+    }
+
+}
+impl<EM, I, OT, S> Executor<I> for ForkserverExecutor<EM, I, OT, S>
+where
+    I: Input + HasTargetBytes,
+    OT: ObserversTuple,
+{
+    #[inline]
+    fn run_target(&mut self, _input: &I) -> Result<ExitKind, Error> {
+
+        let slen = self.forkserver.write_ctl(self.forkserver().last_run_timed_out());
+        if slen != 4{
+            panic!("failed to request new process");
+        }
+
+        let (rlen, pid) = self.forkserver.read_st();
+        if rlen != 4 {
+            panic!("failed to request new process")
+        }
+        if pid <= 0 {
+            panic!("forkserver is misbehaving");
+        }
+
+        //child running
+        let (_, status) = self.forkserver.read_st();
+        self.forkserver.status = status;
+
+        Ok(ExitKind::Ok)
+    }
+}
+
+impl<EM, I, OT, S, Z> HasExecHooks<EM, I, S, Z> for ForkserverExecutor<EM, I, OT, S>
+where
+    I: Input + HasTargetBytes,
+    OT: ObserversTuple,
+{
+    #[inline]
+    fn pre_exec(&mut self, _fuzzer:&mut Z, _state: &mut S, _event_mgr: &mut EM, input: &I) -> Result<(), Error>{
+        //write to test case
+        self.out_file.write_buf(&input.target_bytes().as_slice().to_vec());
+        //outfile gets automatically closed.
+        Ok(())
+    }
+
+    fn post_exec(&mut self, _fuzzer:&mut Z, _state: &mut S, _event_mgr: &mut EM, _input: &I) -> Result<(), Error>{
+
+        if !libc::WIFSTOPPED(self.forkserver.status()) {
+            self.forkserver.child_pid = 0;
+        }
+
+        //move the head back
+        self.out_file.rewind();
+
+        if(libc::WIFSIGNALED(self.forkserver.status())){
+            println!("CRASH");
+        }
+        println!("OK");
+        Ok(())
+    }
+}
+
+impl<EM, I, OT, S> HasObservers<OT> for ForkserverExecutor<EM, I, OT, S>
+where
+    I: Input + HasTargetBytes,
+    OT: ObserversTuple,
+{
+    #[inline]
+    fn observers(&self) -> &OT {
+        &self.observers
+    }
+
+    #[inline]
+    fn observers_mut(&mut self) -> &mut OT {
+        &mut self.observers
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use crate::executors::{OutFile, ForkserverExecutor, Forkserver, Executor};
+    use crate::inputs::NopInput;
+    #[test]
+    fn test_forkserver() {
+        let command = "/home/toka/work/aflsimple/test";
```

Comment:
Would be cool to have a testcase, but this one can't work, sadly :)

Reply:
Yes 😄, I think I'll add a simple example binary file and a c source code along with a example fuzzer.

Reply:
I think we should add a `libafl-tests` project that can have more complex testcases, such as a short fuzz session, etc.
Committing binaries will break on various platforms, and in various colors. :)